src/overlays/actors/ovl_En_Fd/z_en_fd.c

```diff
@@ -21,61 +41,796 @@ const ActorInit En_Fd_InitVars = {
+    }
+}
+
+void EnFd_SpinAndSpawnFire(EnFd* this, GlobalContext* globalCtx) {
+    f32 deceleration;
+    f32 tgtSpeed;
+    f32 rotSpeed;
+
+    if ((this->spinTimer < 31) && (this->invincibilityTimer == 0)) {
+        func_8002F974(&this->actor, NA_SE_EN_FLAME_FIRE_ATTACK - SFX_FLAG);
+    } else {
+        func_8002F974(&this->actor, NA_SE_EN_FLAME_ROLL - SFX_FLAG);
+    }
+
+    if (DECR(this->spinTimer) != 0) {
+        this->actor.shape.rot.y += (this->runDir * 0x2000);
+        if (this->spinTimer == 30 && this->invincibilityTimer == 0) {
+            if (this->actor.xzDistFromLink > 160.0f) {
+                // orange flames
+                EnFd_SpawnChildFire(this, globalCtx, 8, 0);
+            } else {
+                // blue flames
+                EnFd_SpawnChildFire(this, globalCtx, 8, 1);
+            }
+        }
+    } else {
+        // slow shape rotation down to meet `this` rotation within ~1.66 degrees
+        deceleration = this->actor.posRot.rot.y;
+        deceleration -= this->actor.shape.rot.y;
+        rotSpeed = 0.0f;
+        tgtSpeed = fabsf(deceleration);
+        deceleration /= tgtSpeed;
+        Math_SmoothScaleMaxF(&rotSpeed, tgtSpeed, 0.6f, 8192.0f);
+        rotSpeed *= deceleration;
+        this->actor.shape.rot.y += (s16)rotSpeed;
+        rotSpeed = fabsf(rotSpeed);
+        if ((s32)rotSpeed < 0x12D) {
+            // ~1.6 degrees
+            this->actor.shape.rot.y = this->actor.posRot.rot.y;
+        }
+
+        if (this->actor.shape.rot.y == this->actor.posRot.rot.y) {
+            this->initYawToInitPos = Math_Vec3f_Yaw(&this->actor.initPosRot.pos, &this->actor.posRot.pos);
+            this->curYawToInitPos = this->runDir < 0 ? 0xFFFF : 0;
+            this->circlesToComplete = (globalCtx->state.frames & 7) + 2;
+            this->spinTimer = Math_Rand_S16Offset(30, 120);
+            func_80034EC0(&this->skelAnime, &sAnimations, 3);
+            this->actionFunc = EnFd_Run;
+        }
+    }
+}
+
+/**
+ * Run around in a circle with the center being the initial position, and
+ * the radius being the distance from the initial postion to the nearest
+ * threat (bomb or player).
+ */
+void EnFd_Run(EnFd* this, GlobalContext* globalCtx) {
+    Actor* potentialThreat;
+    s16 yawToYawTarget;
+    f32 runRadiusTarget;
+    Vec3f adjPos;
+
+    if (EnFd_ShouldStopRunning(this, globalCtx, this->runRadius, &this->runDir)) {
+        if (this->invincibilityTimer == 0) {
+            this->actor.posRot.rot.y ^= 0x8000;
+            this->actor.velocity.y = 6.0f;
+            this->actor.speedXZ = 0.0f;
+            func_80034EC0(&this->skelAnime, &sAnimations, 1);
+            this->actionFunc = EnFd_JumpToGround;
+            return;
+        }
+    }
+
+    yawToYawTarget = Math_Vec3f_Yaw(&this->actor.initPosRot.pos, &this->actor.posRot.pos) - this->initYawToInitPos;
+    if (this->runDir > 0) {
+        if ((u16)this->curYawToInitPos > (yawToYawTarget & 0xFFFF)) {
```

Comment:
Can these be written as (u16) casts? Is `yawToYawTarget` possibly a u16? Maybe `this->curYawToInitPos` is too?